### DISCUSSION TITLE
Model fine-tuning, rare object

### DISCUSSION THREAD

**Discussion by MuradOmarov** [ultralytics/ultralytics#21360] [Q&A]
Dear Ultralytics community and developers,

First of all, thank you for the amazing workflow you've developed.

I have a question regarding model fine‑tuning.

I previously trained a YOLOv8 model that works reasonably well for single‑object detection on grayscale medical images. For that project, I used only 800 images, which was enough.

Now I have data with the same object but at a slightly different resolution and zoom level. I’d like to fine‑tune the previously trained model to detect the object in this new dataset. Note that the prevalence of the object has changed from 35 % to 28 % (the object is present in only 28 % of the images).

I’ve tried various strategies—such as freezing the first 10 layers and lowering the learning rate—and I’m currently experimenting with a two‑phase fine‑tuning approach

```
model = YOLO('old_model.pt')

model.train(data = '/content/drive/MyDrive/new_dataset.yaml',
                      epochs=10, patience = 5, seed = 8, lr0=0.0001, 
                      **freeze = 10**, 
                       device =DEVICE, name ='Freeze', project ='FineTune', batch = 44, single_cls=True, 
                      cos_lr = True,
                      optimizer='SGD', pretrained=True,
                      imgsz=480)

model = YOLO('FineTune/Freeze/weights/best.pt')

model.train(data = '/content/drive/MyDrive/new_dataset.yaml',
                      epochs=35, patience = 5, seed = 8,l r0=0.0001, device =DEVICE, name ='resume', project ='FineTune',
                      batch = 44,single_cls=True, cos_lr = True,
                      optimizer='SGD', pretrained=True, imgsz=480,
                      # Augmentations
                     mosaic = 0,
                     fliplr = 0.1, perspective = 0, copy_paste=0,
                      hsv_s=0.05,hsv_v=0.05,hsv_h = 0.05,mixup = 0, translate=0.1, scale = 0.1,
                      flipud = 0, degrees= 10,shear=0,close_mosaic = 0
                      )
```

Unfortunately, during the second phase the model seems to overfit, as I infer from the training report (see screenshot).

<img width="876" height="441" alt="Screenshot 2025-07-12 at 16 59 22" src="https://github.com/user-attachments/assets/172c6448-e2d3-4218-90b9-ca7173db93bc" />

When I use only the single‑phase approach with 10 frozen layers, the mAP@50 doesn’t rise above 0.48, whereas it was 0.685 on the original dataset.

Do you have any suggestions on which strategy I should try next?

Thank you in advance.

Best wishes,
Murad

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MuradOmarov, thank you for your detailed description and for using Ultralytics for your medical imaging workflow! 🚀 Your systematic approach to fine-tuning and your experiment logs are very helpful for understanding the context.

If this is a 🐛 Bug Report, please also provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

If you're seeking fine-tuning advice or help with rare object detection, please include as much relevant information as possible—such as example images, dataset stats, or specific training logs. For custom training, make sure to review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

We encourage you to check out our [Docs](https://docs.ultralytics.com/), especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) sections, as many common fine-tuning scenarios are covered there.

Connect with the Ultralytics community in the way that suits you best:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth Q&A
- [Subreddit](https://www.reddit.com/r/Ultralytics/) for sharing experiences and advice

## Upgrade

To ensure you benefit from the latest features and bug fixes, upgrade to the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be trained and evaluated in these up-to-date environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing, verifying correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

This is an automated response 🤖—an Ultralytics engineer will review your discussion and offer further assistance soon!

**Reply by glenn-jocher** [reply]
Hi @MuradOmarov! Your two-phase fine-tuning approach is sound, but the overfitting in the second phase suggests a few adjustments could help.

For rare object detection scenarios like yours, consider setting `warmup_epochs=0` to start with a higher learning rate immediately since your model already has learned features from the previous dataset. You might also try reducing the learning rate further in the second phase (e.g., `lr0=0.00005`) and adding stronger regularization with `weight_decay=0.0005`.

Given the reduced object prevalence (35% to 28%), you could benefit from [image tiling techniques](https://docs.ultralytics.com/guides/model-evaluation-insights/#image-tiling-for-small-objects) if your objects are small, and consider class balancing strategies or focal loss to handle the imbalanced dataset better.

Since you're working with medical imaging, I'd also recommend trying YOLO11 which offers improved performance over YOLOv8 - simply replace `YOLO('old_model.pt')` with a YOLO11 model and fine-tune from there.

For the overfitting issue, try reducing the number of epochs in the second phase or implementing early stopping with a validation set to monitor when performance starts degrading.

**Reply by MuradOmarov** [reply]
Thank you very much!

This is how the second phase look like with lr0=0.00005, warmup_epochs=0 and weight_decay=0.0005:

<img width="521" height="436" alt="Screenshot 2025-07-13 at 18 23 11" src="https://github.com/user-attachments/assets/bd78cdaf-c4cd-4345-9871-b3abb7f58b41" />


I was wondering what the best way would be to apply the class‑balancing strategies or the focal loss you mentioned above. Would a GridSearch be appropriate?

For example:

param_grid = {
    "box": [3, 6.5, 10],
    "cls": [0.15, 0.5, 0.8, 1.1],
    "dfl": [0.5, 1.5, 2.5],
}

And also should implement a grid search for the first or second phase of fine-tuning? 

Thank you in advance!

**Reply by glenn-jocher** [reply]
Great progress with those hyperparameter adjustments! For class-balancing and loss function tuning, I'd recommend using Ultralytics' built-in [hyperparameter tuning system](https://docs.ultralytics.com/guides/hyperparameter-tuning/) rather than manual GridSearch.

You can use the `model.tune()` method which is more efficient and designed specifically for YOLO models:

```python
# After your first phase training
model = YOLO('FineTune/Freeze/weights/best.pt')
results = model.tune(data='/content/drive/MyDrive/new_dataset.yaml', iterations=10, imgsz=480)
```

This will automatically optimize hyperparameters including the loss weights (box, cls, dfl) you mentioned. I'd recommend applying this tuning during the **second phase** since you want to optimize for your specific new dataset after the initial frozen layer training.

For rare object detection, the `cls` weight is particularly important - higher values (like 1.1-1.5) can help with class imbalance. The tuning system will find the optimal balance automatically while considering your specific dataset characteristics.

This approach is more systematic than manual GridSearch and will likely yield better results for your medical imaging use case.

**Reply by MuradOmarov** [reply]
Dear Glenn,

Very insightful! 
Thank you very much for your help!

**Reply by glenn-jocher** [reply]
You're very welcome @MuradOmarov! I'm glad the hyperparameter tuning approach will be helpful for your medical imaging project. 

Feel free to share your results or reach out if you encounter any issues during the tuning process. Best of luck with your rare object detection work!
